### DISCUSSION TITLE
Memory leak in Celery Beat 5.3.4 process

### DISCUSSION THREAD

**Discussion by chuck-white-willo** [celery/celery#8645] [Q&A]
As others seem to be reporting here, there is a pretty obvious memory leak in Beat 5.3.4.  I am running on python 3.8.16 and Django  4.2.6.  We see this across a number of our apps, and when running on a local dev environment.  It's very easy to re-create: just start up Django via gunicorn and then start celery beat using the django database scheduler.  If you watch the memory allocated to the beat process, it sits there and just keeps climbing.  I only have 1 periodic task in django that literally just returns "OK" every 1 minute, and even without that, the memory usage of beat constantly climbs.

We also have a celery 4.4.2 environment where this isn't an issue.

I am running memray on my local now to capture.  But is that really needed if this is so easily reproducible?

**Comment by auvipy** [comment]
can you use memray and track down the root of memory leaking codes?

**Comment by chuck-white-willo** [comment]
@auvipy  I had a very hard time getting Memray to both start and attach to the the process with Memray's cmdline syntax and all the args we send to our beat process.  But I was able to attach to an existing running beat process after much tinkering.  So you will see in the screenshot that the total heap is only about 4MB as that is what grew in a couple of hours after starting the process.  Actual memory allocation on the host is 87MB resident.

Throughout the entire trace period `psycopg2/__init__.py` continually grows in size.  We are running version 2.9.9 of psycopg2.  I confirmed that no database connections are being left open.  The only periodic task running is a very simple healthcheck, once per minute, that simply returns 'OK'.

![Screenshot 2023-11-16 at 2 12 18 PM](https://github.com/celery/celery/assets/133787913/12562e36-76bc-4b04-ae62-00694da210db)

**Comment by chuck-white-willo** [comment]
I let that same live trace run overnight.  You can see the leak progression over time.  It's slow, but eventually the host hits 100% mem usage after many days.

![Screenshot 2023-11-17 at 10 09 39 AM](https://github.com/celery/celery/assets/133787913/f71974c4-653b-43ec-8458-19ba68e59d96)

**Comment by chuck-white-willo** [comment]
@auvipy I let that trace run all weekend.  It's slow, but it keeps climbing.  Any thoughts on how to mitigate this?
![Screenshot 2023-11-20 at 9 11 30 AM](https://github.com/celery/celery/assets/133787913/f8002cce-ddbf-4413-b63b-7c176f6e8963)
